### DISCUSSION TITLE
Incompatibility - C# allows optional+params, VB does not

### DISCUSSION THREAD

**Discussion by mattjohnsonpint** [dotnet/csharplang#7096] [Q&A]
Consider a class library that has an interface with a method taking an optional parameter and a `params` array:

```csharp
public interface ISomething
{
    void DoSomething(int foo = 0, params string[] bar);
}
```

This works fine in C#.  But a consuming application might be written in VB.NET, and there's no way to implement this interface.

One would think it would be:

```vb
Public Class Something
    Implements ISomething

    Public Sub DoSomething(Optional foo As Integer = 0, ParamArray bar As String()) Implements ISomething.DoSomething
    End Sub
End Class
```

But this gives a compile error:

```
[BC30046] Method cannot have both a ParamArray and Optional parameters.
```

Removing either `Optional` or `ParamArray` from the VB definition doesn't work, because it no longer conforms to the interface.

How can one prevent such problems?  I'm not sure that VB should support this, but I'd at least like to be warned when compiling C# code that some consumers might not be able to implement the interface.  I've set `[assembly: CLSCompliant(true)]` on the C# library because I thought that's exactly what is for - but it does not warn or error on this.

Is such code CLS compliant?
- If yes, why doesn't VB allow it?
- If no, is this a compiler error with CLS Compliance checking?

Should an analyzer have found this?

Also, is this the best place to discuss this issue?

Thanks

**Comment by HaloFour** [comment]
I think CLS compliance only checks the individual features, and optional parameters and params are each separately CLS.

Really makes me wonder how C# 1-3 would have handled implementing a VB interface that declared a method with optional parameters.

**Comment by tannergooding** [comment]
You might be able to make this work by manually specifying the relevant attributes that allow `optional` to work (both `OptionalAttribute` and `DefaultParameterValueAttribute` are defined in `System.Runtime.InteropServices`):
```
Public Sub DoSomething(<[Optional], DefaultParameterValue(0)> foo As Integer, ParamArray bar As String()) Implements ISomething.DoSomething
```

Such definitions should be "correctly" encoded in IL, and therefore consumable "as expected" from languages that support such a combination. But, it may still run into quirks on the VB side of things.

**Reply by HaloFour** [reply]
Unfortunately VB.NET does not recognize the combination of attributes as being equivalent to the language feature.

IIRC, it's not the IL that's the problem here.  That metadata isn't part of the signature.  C# doesn't require it, nor does the C# compiler emit it into the IL:  [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIDMABLAgKYBOAZhAMZlECSRA3gLABQR3ReALEQCyAClIkiAXiIAGOEQAOEChDABnXgEZpAbQC6RJQHNVASgDcnAL6dOhXgCYiAYSIgmrTj17F+Q0TAQSOTwtPQMKYxNWImsOSyA)

**Reply by mattjohnsonpint** [reply]
Yeah, @tannergooding that doesn't compile.

```
error BC30401: 'DoSomething' cannot implement 'DoSomething' because there is no matching sub on interface 'ISomething'.
error BC30149: Class 'Something' must implement 'Sub DoSomething([foo As Integer = 0], ParamArray bar As String())' for interface 'ISomething'.
```

And my IDE (Rider) gives:

<img width="1370" alt="image" src="https://user-images.githubusercontent.com/1396388/230633155-661726ce-ba7e-42fd-a539-4619352930c8.png">

**Reply by tannergooding** [reply]
Ah, that's unfortunate. I didn't realize VB also required the optional parameters to match.

**Reply by mattjohnsonpint** [reply]
It's really not a good API design anyway.  Regardless of the implementation issues, without the optional first parameter provided, one has to use named parameters to get at the second, just to invoke the method.

This works in C#

```csharp
public static void Whatever(ISomething something)
{
    something.DoSomething(bar: "abc");
    something.DoSomething(bar: new[] {"abc", "def"});
}
```

But the equivalent fails in VB

```vb
Public Shared Sub Whatever(something As ISomething)
    something.DoSomething(bar := "abc")
    something.DoSomething(bar := new String() {"abc", "def"})
End Sub
```
```
[BC30587] Named argument cannot match a ParamArray parameter.
```

So it's not that I want a way to make it work in VB, but rather I want the C# compiler (or an analyzer) to warn about making such an API in the first place.

**Reply by mattjohnsonpint** [reply]
Oh, but I guess VB lets you call it like this:

```vb
Public Shared Sub Whatever(something As ISomething)
    something.DoSomething(, "abc")
    something.DoSomething(, "abc", "def")
    something.DoSomething(, new String() {"abc", "def"})
End Sub
```

who knew!  😅
